### PR TITLE
MWPW-152082 - Utils to Scripts

### PR DESCRIPTION
**Pull request by JasonHowellSlavin**
* Moves set/getLibs from utils into scripts to remove dependency for performance reasons. 

To Test:
When testing, please make sure you do the following at least: 

- test locally with `aem up`
- test the branch url (though it is in the "after")
- test by running milo libs locally, and using milo libs on our branch

Resolves: [MWPW-152082](https://jira.corp.adobe.com/browse/MWPW-152082)

**Test URLs:**
- Before: https://main--bacom--adobecom.hlx.live/?martech=off
- After: https://utils-to-scripts--bacom--adobecom.hlx.live/?martech=off

**Verification URLS for Consumer Blocks**

https://utils-to-scripts--bacom--adobecom.hlx.page/docs/library/blocks/chat-cta
https://utils-to-scripts--bacom--adobecom.hlx.page/docs/library/blocks/comparison-table
https://utils-to-scripts--bacom--adobecom.hlx.page/docs/library/blocks/event-speakers
https://utils-to-scripts--bacom--adobecom.hlx.page/docs/library/blocks/stats
https://utils-to-scripts--bacom--adobecom.hlx.page/docs/library/blocks/tree-view
https://utils-to-scripts--bacom--adobecom.hlx.page/docs/library/blocks/workfront-login

### CODE REVIEW

**Review by meganthecoder — COMMENTED** [review]
Is this the only reference to set/getlibs? Don't we use them in tests too?

**Comment by JasonHowellSlavin** [comment]
> Is this the only reference to set/getlibs? Don't we use them in tests too?

We do use `set/getLibs` in tests and other files which is why I left it as an export in the `utils.js` file and did not delete `utils.js`. 

I don't think the removal of the import from scripts ought to effect anything, but Dennis should probably run a full smoke test.

**Comment by meganthecoder** [comment]
> > Is this the only reference to set/getlibs? Don't we use them in tests too?
> 
> We do use `set/getLibs` in tests and other files which is why I left it as an export in the `utils.js` file and did not delete `utils.js`.
> 
> I don't think the removal of the import from scripts ought to effect anything, but Dennis should probably run a full smoke test.

Oh yeah. I didn't notice that you hadn't deleted utils.js. I assumed you were moving the function, not duplicating it. I don't understand how that works. Isn't `libs` scoped to that function, not on the window? I don't see how you can set libs in scripts.js and get it in utils.js or am I missing something...

**Comment by JasonHowellSlavin** [comment]
> Oh yeah. I didn't notice that you hadn't deleted utils.js. I assumed you were moving the function, not duplicating it. I don't understand how that works. Isn't `libs` scoped to that function, not on the window? I don't see how you can set libs in scripts.js and get it in utils.js or am I missing something...

In this instance, not so sure the closure matters since the function runs immediately and returns the get/set pattern and scripts looks for the global of `window.location` if a location is not provided to the first arrow function. We actually don't need libs for our `utils.js`, and we definitely don't need to make the call for our marquees. Before copying the libs pattern into scripts, we were essentially saying "scripts, go get bacom utils to figure out which path to load milo utils", now we're saying "scripts use setLibs which you have to figure out which utils to get". Were' cutting out the utils call here that blocks milo utils from running loadArea etc.

Screen Shot 2024-06-20 at 9 15 54 AM
Screen Shot 2024-06-20 at 9 31 33 AM

I left the function in utils because quite a bit of our code uses `getLibs` in order to use Milo core functionality: https://github.com/search?q=repo%3Aadobecom%2Fbacom%20getLibs&type=code

File: scripts/scripts.js [hunk 1]
```diff
@@ -157,6 +155,22 @@ const loadStyle = (path) => {
   eagerLoad(marquee.querySelector('img'));
 }());
 
+export const [setLibs, getLibs] = (() => {
```
Comment (Brandon32) [inline scripts/scripts.js:158]: Do we have to export here? A little concerned about side effects from self-executing functions if someone actually imported it.

File: scripts/scripts.js [hunk 2]
```diff
@@ -157,6 +155,22 @@ const loadStyle = (path) => {
...
+    (prodLibs, location) => {
+      libs = (() => {
+        const { hostname, search } = location || window.location;
+        if (!(hostname.includes('.hlx.') || hostname.includes('local'))) return prodLibs;
+        const branch = new URLSearchParams(search).get('milolibs') || 'main';
+        if (branch === 'local') return 'http://localhost:6456/libs';
+        return branch.includes('--') ? `https://${branch}.hlx.live/libs` : `https://${branch}--milo--adobecom.hlx.live/libs`;
+      })();
+      return libs;
+    }, () => libs,
```
Comment (Brandon32) [inline scripts/scripts.js:170]: `getLibs` is a little redundant in this context.

**Review by Brandon32 — APPROVED** [review]

**Comment by meganthecoder** [comment]
@JasonHowellSlavin Shouldn't this be working? https://utils-to-scripts--bacom--adobecom.hlx.page/drafts/methomas/tree-view

File: scripts/scripts.js [hunk 3]
```diff
@@ -157,18 +154,25 @@ const loadStyle = (path) => {
   eagerLoad(marquee.querySelector('img'));
 }());
 
-const miloLibs = setLibs(LIBS);
+// eslint-disable-next-line import/prefer-default-export
+export const LIBS = (() => {
+  const { hostname, search } = window.location;
+  if (!(hostname.includes('.hlx.') || hostname.includes('local'))) return LIBS;
```
Comment (meganthecoder) [inline scripts/scripts.js:160]: This line is different than what we have in utils:
`if (!['.hlx.', '.stage.', 'local'].some((i) => hostname.includes(i))) return prodLibs;`

**Comment by meganthecoder** [comment]
Also, what about the tests? If we don't need to set the libs anymore, can we remove that from the tests?

File: scripts/scripts.js (hunk 3 above)
Comment (Brandon32) [inline scripts/scripts.js:160]: `LIBS` returns `LIBS`? Should this be '/libs'?

**Comment by Brandon32** [comment]
> Also, what about the tests? If we don't need to set the libs anymore, can we remove that from the tests?

We need some tests here, this is too important not to test.

**Comment by meganthecoder** [comment]
> > Also, what about the tests? If we don't need to set the libs anymore, can we remove that from the tests?
> 
> We need some tests here, this is too important not to test.

Agreed, but I think we're talking about two different things:

1. Like you said, we need unit tests for this. Maybe the tests from `utils.test.js` can be moved/adapted.
2. I'd also like us to clean up the block tests like `tree-view.test.js` that import and use `setLibs` from utils. If we're no longer using `setLibs` in the block code, then we don't need to have in the block tests, right?

**Comment by JasonHowellSlavin** [comment]
> > > Also, what about the tests? If we don't need to set the libs anymore, can we remove that from the tests?
> > 
> > 
> > We need some tests here, this is too important not to test.
> 
> Agreed, but I think we're talking about two different things:
> 
> 1. Like you said, we need unit tests for this. Maybe the tests from `utils.test.js` can be moved/adapted.
> 2. I'd also like us to clean up the block tests like `tree-view.test.js` that import and use `setLibs` from utils. If we're no longer using `setLibs` in the block code, then we don't need to have in the block tests, right?

@Brandon32 , I will adapt the `unit.test.js` for this use case as Megan is saying, good point both. 

For point 2. I think we need to look into that issue separately from this. There is some flakiness in our tests as they stand now, we may need an issue to look into them and refactor a bit. Right now many of the tests are throwing errors into the Browser logs, but are still passing, like tree-view, redirects-formatter, etc.

**Comment by JasonHowellSlavin** [comment]
After discussing with team, and in light of my absence for a few weeks, we are closing this PR in favor of https://github.com/adobecom/bacom/pull/220.

**Pull request closed without merging**